editor-service/src/main/webapp/app/dsm/diagram/controller/PaletteDiagramEditorController.ts

```diff
@@ -0,0 +1,156 @@
+/// <reference path="../../../common/interfaces/editorCore.d.ts" />
+/// <reference path="../../../common/interfaces/vendor.d.ts" />
+/// <reference path="../../../../resources/thrift/editor/PaletteService_types.d.ts" />
+/// <reference path="../../../../resources/thrift/editor/PaletteServiceThrift.d.ts" />
+/// <reference path="../../../../resources/thrift/struct/Palette_types.d.ts" />
+/// <reference path="../../../../resources/types/thrift/Thrift.d.ts" />
+/// <reference path="../../../common/constants/GeneralConstants.ts" />
+/// <reference path="../exporters/PaletteExporter.ts" />
+/// <reference path="../parsers/PaletteParser.ts" />
+/// <reference path="../model/PaletteView.ts" />
+
+class PaletteDiagramEditorController extends DiagramEditorController {
+
+    private exporter: PaletteExporter;
+    private parser: PaletteParser;
+    private palettes: PaletteView[] = [];
+    private meta: boolean = true;
+
+    constructor($scope, $attrs) {
+        super($scope, $attrs);
+
+        this.exporter = new PaletteExporter();
+        this.parser = new PaletteParser(this);
+
+        $scope.createPalette = () => { this.createPalette(); };
+        $scope.openPalettesMenu = () => { this.openPalettesMenu(); };
+        $scope.loadMetaEditor = () => { this.loadMetaEditor(); };
+        this.elementsTypeLoader.load((elementTypes: ElementTypes): void => {
+            this.handleLoadedTypes(elementTypes);
+        }, "dsm");
+
+        var controller = this;
+        try {
+            var tPalettes: TPaletteView[] = controller.getClient().getPalettes();
+            for (var i = 0; i < tPalettes.length; i++) {
+                controller.palettes.push(PaletteView.createFromDAO(tPalettes[i]));
+            }
+        }
+        catch (e) {
+            console.log("Error: can't get folder tree", e);
+        }
+    }
+
+    public handleLoadedTypes(elementTypes: ElementTypes): void {
+        this.propertyEditorController = new PropertyEditorController(this.sceneController, this.undoRedoController);
+
+        for (var typeName in elementTypes.uncategorisedTypes) {
+            this.nodeTypesMap[typeName] = elementTypes.uncategorisedTypes[typeName];
+        }
+
+        var categories: Map<Map<NodeType>> = elementTypes.paletteTypes.categories;
+        for (var category in categories) {
+            for (var typeName in categories[category]) {
+                this.nodeTypesMap[typeName] = categories[category][typeName];
+            }
+        }
+
+        this.paletteController.appendBlocksPalette(elementTypes.paletteTypes);
+        this.paletteController.initDraggable();
+    }
+
+    public createPalette() {
+        if (this.meta) {
+            var name: string = prompt("input palette name");
+            if (name !== "" && name !== null) {
```

Comment:
the order of the checks looks weird. it checks for the variable's value and **then** checks whether it is null